### ISSUE TITLE
How does overzooming work?

### ISSUE THREAD

**Issue by mapsam** [mapbox/vector-tiles#21]
[Question from @jczaplew](https://twitter.com/johnjcz/status/765576636582420480) 😄 

> How does overzooming [at] certain scales work? i.e. tiles for z5 are valid to z8, z9 are valid to z11, etc..

**Comment by mapsam** [comment]
Awesome question @jczaplew. Going to answer this question from the beginning for others reading in.

## What is overzooming? 

Overzooming is a vector-tile specific technique that allows a specific tile to be rendered beyond its intended zoom level, so it continues to be visible on the map. If a tileset had a `minzoom` of 6 and a `maxzoom` of 12, those are the valid ranges calculated by the tile generator (more on that below). If you were to zoom your map beyond zoom level 12, the map renderer can continue to use zoom 12 by scaling the vector data upwards. This, of course, can result in gross simplifications of data if geometries are rendered too far above their actual level of detail.

## How is _zoom range_ calculated?

In short, vector tilesets use filesize to help decide zoom range. Raster tilesets use pixel resolution to generate min/max zoom levels. An example of a tool that calculates these ranges is [Mapnik](https://github.com/mapnik/mapnik) or, subsequently, [mapnik omnivore](https://github.com/mapbox/mapnik-omnivore/). The actual logic here is beyond my understanding, but maybe @springmeyer could say a few words about how Mapnik calculates these zooms?

## How is _overzoom_ calculated?

In vector tiles overzoom is the combination between two different min/max zoom parameters:

1. the tileset itself has vector tiles generated between two zooms (min and max) and these are stored in a filesystem somewhere according to the zoom level directory. Example min/max for the tiles generated: `10/14`.
1. the [style](https://www.mapbox.com/mapbox-gl-style-spec/) can specify a [min and max zoom](https://www.mapbox.com/mapbox-gl-style-spec/#sources) for a tileset "layer". This style can specify min/max zooms that are different from the actual tileset in 1. An example of a style specifying min/max zoom for the `landuse` layer: `10/17`. 

Notice how the style in 2 has a maxzoom of `17`, which is larger than the max `14` generated in the tiles themselves? This is the range where overzooming occurs. So for zoom levels `15`, `16`, and `17` - the z14 tile is scaled upwards by the renderer.

*Note: tiles are generated for EVERY zoom between the min and max specified in 1. The original question seems to imply that overzooming can happen in multiple places for a tileset, (i.e. zoom 10 is good through zoom 13, then zoom 14 is good through zoom 17, etc.). This is not the case, as every tile between 13 and 17 is generated and requested.*

## How is overzooming implemented by a mapping library?

In mapbox-gl-js, overzooming acts similar to a standard tile request, by sending the request to a worker. The worker will re-read the vector source already rendered and apply a scale/transform/translation to redraw it accordingly. The renderer must know that a tile, say zoom 15, does not exist over the wire (i.e. gets a 404 when requesting it) and recognize that the style says to continue styling this layer anyways, thus calculating a scaled version of the z14 tile.

Tiles are redrawn on _every_ frame. If the tile happens to be overzoomed, the transformation matrix is constructed to make sure things are placed correctly. [Here's where that happens in mapbox-gl-js](https://github.com/mapbox/mapbox-gl-js/blob/527213cc92285ca8301a065fd94322ee02bb6e42/js/geo/transform.js#L360-L361). 

**Raster overzooming**: When scaling a raster image (i.e. between z levels like `5.5`) they are "overzoomed" or _scaled_ to the next zoom in order to retain a complete image.

*Note: this becomes a bit tricky when dealing with "label collisions" during overzooming. Since collision points will change as the tile is scaled up, they must be recalculated, which requires the entire tile to be reprocessed.*

cc @lucaswoj @GretaCB @rclark @springmeyer @mourner @lyzidiamond if they have any corrections here.

**Comment by lucaswoj** [comment]
> The renderer must know that a tile, say zoom 15, does not exist over the wire (i.e. gets a 404 when requesting it) and recognize that the style says to continue styling this layer anyways, thus calculating a scaled version of the z14 tile.

Vector tile sources communicate the max zoom at which they provide tiles via [TileJSON](https://github.com/mapbox/tilejson-spec/blob/master/2.0.0/README.md).

**Comment by jczaplew** [comment]
Thank you for the awesome description Sam! 

In "How is overzoom calculated", you give an example where tiles only exist from z10-14, but they are valid all the way up to z17. Presumably, when a user lands on z12, and zooms in to z17 the mapping library will request tiles for z14+, but noticing that they don't exist, continue to use the data from z14. However, how does the mapping library handle a situation in which the user loads the map at, say, z17?

**Comment by mapsam** [comment]
@jczaplew I believe when a tile (or tile source) is loaded into the map, the TileJSON min/max and Style min/max are checked even before it draws, so the initial tile would render as overzoomed, meaning that overzooming isn't necessarily tied to user gestures or zoom animations, but rather zoom state.

**Comment by jczaplew** [comment]
Hypothetical situation:

Let's say I want to make a map, and there are three scales at which I want show completely different data (call them small, medium, and large). The data in each scale is completely independent of the other scales, and I want z0-z10. Let's also assume that small is z0-2, medium is z3-6, and large is z7-10.

Traditionally, I would have to create tiles for all zoom levels between z0 and z10, but it seems like one could be clever with vector tiles to save bandwidth. My understanding is that if the data at z7 and z8 is exactly the same, there is no reason to create tiles for _both_ z7 and z8, and instead we could rely on overzooming. 

With this scenario, would it make sense to create a separate tileset for each scale _only_ at the minimum zoom level for each scale? I.E. we would roll tiles for small at z0 only, z3 only for medium, and z7 for large. In theory it seems like this would be far more efficient for both the client and the server. Are there potential pitfalls to this approach?

**Comment by flippmoke** [comment]
@jczaplew Short answer, it really depends on the data and how the data is used. It is quite possible that it might be nice and save bandwidth. I mean in GL styles with seamless transitions it is effectively overzooming as you scroll the mouse (having partial zooms). However, it might also not but a savings of bandwidth if the user doesn't zoom in often or at all because you might have more detail at a higher zoom then required. 

Basically, my advice would always be to test and measure.

**Comment by jczaplew** [comment]
> This, of course, can result in gross simplifications of data if geometries are rendered too far above their actual level of detail.

I'm experimenting a little bit with overzooming, and have noticed that I am indeed running into gross simplifications of data. However, it is difficult to determine if A) mapnik simplifies geometries when creating vector tiles, B) Mapbox GL JS simplifies geometries from vector tiles, or C) some combination of those two.

Either way, if I create a tile for a dataset at z1 and overzoom into z4 it is quite obvious that a lot of simplification of the original data occurred somewhere. Any insights? Thanks for all the info! This has been incredibly helpful for understanding vector tiles.

**Comment by mapsam** [comment]
I can't speak directly to Mapbox GL JS, as I'm not very familiar with the rendering side of things, but there is plenty of simplification that goes on in mapnik, via [mapnik-vector-tile](https://github.com/mapbox/mapnik-vector-tile). 

All coordinates in a vector tile (a single vector tile) exist on an *integer*-based grid with `0,0` being the origin at the top left of the tile. X is positive to the right, Y is positive down. The simplification occurs when converting geographic coordinates into vector tile coordinates (from floats to integers, essentially). The overzooming of a particular tile will essentially make that grid visual, like this: 

*Zooming in, the points in this tileset are just scaling upwards rather than loading in new tilesets (overzooming). They are only rendered up to z6, so anything beyond (z7+) will be overzoomed.*

![mapbox-overzoomed-points](https://cloud.githubusercontent.com/assets/1943001/17934228/bec9de10-69dc-11e6-848d-5e921df246b1.gif)

*The points exist on a regular grid within the vector tiles.*

<img width="499" alt="screen shot 2016-08-24 at 9 24 46 am" src="https://cloud.githubusercontent.com/assets/1943001/17934219/b6c9227a-69dc-11e6-8509-bd8adff5155e.png">

So, that's the negative side of overzooming. The logic & algorithms for simplifying data into vector tile coordinates exist in mapnik/mapnik-vector-tile.

**Comment by jczaplew** [comment]
Cool, that explains it, and makes a lot of sense. It seems like my desired approach won't work, but because of the simplification Mapnik does there wouldn't be a performance boost with that strategy anyways.

**Comment by mapsam** [comment]
@jczaplew assuming your strategy is something like "have z1 work through z4, z5 work through z8, etc etc" I'd guess you'd need to actually go the reverse: "z1 through z4 uses the z4 tile, z5 through z8 uses the z8 tile, etc etc" but this would require updates to the renderer (Mapbox GL JS). Or you'd have to increase specificity at z1 - where your vector tile grid is more dense than the standard grid used.

**Comment by cmortazavi** [comment]
Are there any examples of how to implement overzooming in OpenLayers?

**Comment by pathmapper** [comment]
Very interesting insights here, thanks!
When retrieving raster tiles from styles via the Mapbox Static API:
https://www.mapbox.com/api-documentation/#retrieve-raster-tiles-from-styles
Is there any difference regarding performance (how fast the raster tiles are delivered) using a style which is using overzooming compared to a style which is only using zoom levels of tilesets the way no overzooming is needed?
Are the raster tiles rendered when they are requested via the Mapbox Static API or are they pre-rendered and cached for all styles?

**Comment by mapsam** [comment]
Hi @pathmapper, let me know if I'm not understanding your questions right.

> Is there any difference regarding performance (how fast the raster tiles are delivered) using a style which is using overzooming compared to a style which is only using zoom levels of tilesets the way no overzooming is needed?

The performance difference here comes down to the devices and mapbox-gl-js or mapbox-gl-native. If you have more zoom levels to offer, that will make more API requests (which could be considered slower). Overzooming happens pretty seamlessly, but you lose out on accuracy (or image quality) if there are no more zoom levels to request.

> Are the raster tiles rendered when they are requested via the Mapbox Static API or are they pre-rendered and cached for all styles?

If you hit the static API at a zoom level greater than your tilesets max zoom, the static API (via mapbox-gl-native) performs the overzooming operation on our servers before responding with an image.

@cmortazavi I'm not too familiar with OpenLayers. Assuming OpenLayers is harnessing OpenGL, perhaps it does it automatically?

**Comment by pathmapper** [comment]
Hi @mapsam, thanks for your answers! My questions are only regarding performance, not about the loss of accuracy when overzooming.

> If you hit the static API at a zoom level greater than your tilesets max zoom, the static API (via mapbox-gl-native) performs the overzooming operation on our servers before responding with an image.

When using the static API (via mapbox-gl-native), the devices performing the overzooming operations are your servers.

Regarding response time of the servers until the image is ready to be delivered: 
Does it takes longer to generate a raster tile (image) when overzooming needs to be processed compared to generating a raster tile where no overzooming operation needs to be processed (because of the used style/vector tileset)?  

In other words:
When the use case is requesting raster tiles via the static API, would it be better to avoid overzooming regarding performance (how fast the tiles could be delivered)?

**Comment by mapsam** [comment]
Hmmm I think we're starting to hit the extent of my knowledge with mapbox-gl-native. Maybe @bsudekum or @mollymerp can fill in the gaps.

> Does it takes longer to generate a raster tile (image) when overzooming needs to be processed compared to generating a raster tile where no overzooming operation needs to be processed 

I'd assume *yes*, purely because oversampling the image is another processing step. Whether that extra operation adds a significant amount of time to the request, I'm not sure!

**Comment by chriswhong** [comment]
So, is underzooming a thing?  I have a situation where there is very dense polygon data (New York City Tax Lots) in vector tiles.  I'd like to have the `minzoom` of the vector tiles be 14 to keep the tile size smaller, but still render them when zoomed out... is this possible?

**Comment by mapsam** [comment]
Great question @chriswhong - underzooming **is not** a thing, at least for gl-native and gl-js, which only begin rendering at the first existing zoom level. Chatted with @springmeyer to get a bit more information as to _why_. 

Consider overzooming, where you are essentially selecting a portion of the maximum zoom level tile to crop and render on your map. Underzooming would require taking mutliple tiles at the minimum zoom level and unioning them together, plus running through some sort of simplification process on the fly (within gl-native or gl-js). There's no way to know _which_ features should be simplified out or which should stay at this point.

Without underzooming, simplification needs to happen on the tile generation side of things. Tippecanoe has a few flags for handing off the simplification which you can [check out here](https://github.com/mapbox/tippecanoe#tile-resolution).

**Comment by millenbop** [comment]
I'm confused :) Is overzooming supported or not? My bug reports keep getting closed :/

https://github.com/mapbox/mapbox-gl-native/issues/9863

**Comment by kevinstadler** [comment]
@millenbop do your lowest-level tiles explicitly specify until what level they can be overzoomed to (see @lucaswoj's comment above)?

**Comment by millenbop** [comment]
@kevinstadler You mean specifying "maxzoom" in the layer in the client?

```
"sources": {
        "composite": {
            "url": "mapbox://mapbox.mapbox-terrain-v2,mapbox.mapbox-streets-v7",
            "type": "vector"
        },
        "mapillary": {
            "tiles": [
                      "https://d25uarhxywzl1j.cloudfront.net/v0.1/{z}/{x}/{y}.mvt"
                      ],
            "type": "vector"
        }
    },
...
 "layers": [
{
            "id": "mapillary-sequence-overview",
            "type": "circle",
            "source": "mapillary",
            "source-layer": "mapillary-sequence-overview",
            "paint": {
               "circle-opacity": 0.6,
               "circle-color": "#39AF64",
               "circle-radius": 4
            }
        },               
        {
            "id": "mapillary-sequences",
            "type": "line",
            "source": "mapillary",
            "source-layer": "mapillary-sequences",
            "interactive": true,
            "layout": {
               "line-join": "round",
               "line-cap": "round"
            },
            "paint": {
               "line-color": "#39AF64",
               "line-width": 2
            }
        },
...
```

**Comment by millenbop** [comment]
I tried adding "maxzoom" but it just stops it from showing, as expected.

**Comment by mapsam** [comment]
Looks like @millenbop's issue was resolved over at https://github.com/mapbox/mapbox-gl-native/issues/9863#issuecomment-325615680 thanks to some info from @kkaefer!

**Comment by scarletjohansson** [comment]
@mapsam in terms of "underzooming", why couldn't the renderer just rendering 4x or 8x or 16x tiles in place of the single tile, without combining or simplifying the geometry into a single tile.  I know it's not an optimal solution but wouldn't it at least work for cases where the zoom levels aren't available (for example realtime data, etc, that would cumbersome to generate at multiple zoom levels).

**Comment by mapsam** [comment]
@scarletjohansson great question! Renderers _could_ implement that if they so desired. As far as I understand, a couple reasons why you _wouldn't_ implement this are:

1. Loading tiles is expensive, so if you're loading 4-8-16x the number of tiles you're making 4-8-16x the requests to whichever tile API you're using. That could be hundreds of requests for a single page map application.
1. Loading tiles into smaller space would require the renderer to make some decisions about which features to render over others. For example, a z4 vector tile rendered at zoom 4 is rendering features at a 1-to-1 scale. If you zoom out to z3 and are rending four z4 tiles in the place of a single z3 tile, you now have a 4-to-1 scale for a single pixel, meaning could be choosing between four possible values from that single z4 vector tile to render in a single pixel at the z3 space. Style and layer ordering/filtering could help here, but what if you have four features all at the same priority level in a style? Which do you render? This wouldn't be an obvious choice.

Perhaps there are more reasons, will share as I learn!

**Comment by stevage** [comment]
I'm also confused about whether it's possible to request overzooming that a given layer be overzoomed in mapbox-gl-js, and if so, how?

**Comment by stevage** [comment]
Oh, ok, it does work. In the vector tile source, set `maxzoom` to the level of tile available, and that will be overzoomed between there and the map's `maxzoom` property.
